bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LedgerMetadataSerDe.java

```diff
@@ -66,59 +76,70 @@
-                s.append(FIELD_SPLITTER);
-                s.append(addr.toString());
+        for (int i = 0; i < VERSION_KEY_BYTES.length; i++) {
+            int b = is.read();
+            if (b < 0 || ((byte) b) != VERSION_KEY_BYTES[i]) {
+                throw new IOException("Ledger metadata header corrupt at index " + i);
             }
         }
-
-        if (metadata.getState() == State.IN_RECOVERY) {
-            s.append(LINE_SPLITTER).append(V1_IN_RECOVERY_ENTRY_ID)
-                .append(FIELD_SPLITTER).append(V1_CLOSED_TAG);
-        } else if (metadata.getState() == State.CLOSED) {
-            s.append(LINE_SPLITTER).append(metadata.getLastEntryId())
-                .append(FIELD_SPLITTER).append(V1_CLOSED_TAG);
-        } else {
-            checkArgument(metadata.getState() == State.OPEN,
-                          String.format("Unknown state %s for V1 serialization", metadata.getState()));
+        byte[] versionBuf = new byte[MAX_VERSION_DIGITS];
+        int i = 0;
+        while (i < MAX_VERSION_DIGITS) {
+            int b = is.read();
+            if (b == LINE_SPLITTER_BYTES[0]) {
+                String versionStr = new String(versionBuf, 0, i, UTF_8);
+                try {
+                    return Integer.parseInt(versionStr);
+                } catch (NumberFormatException nfe) {
+                    throw new IOException("Unable to parse version number from " + versionStr);
+                }
+            } else if (b < 0) {
+                break;
+            } else {
+                versionBuf[i++] = (byte) b;
+            }
         }
+        throw new IOException("Unable to find end of version number, metadata appears corrupt");
+    }
 
+    public byte[] serialize(LedgerMetadata metadata) throws IOException {
+        int formatVersion = Math.min(maxLedgerMetadataFormatVersion, metadata.getMetadataFormatVersion());
+        final byte[] serialized;
+        switch (formatVersion) {
+        case 3:
+            serialized = serializeVersion3(metadata);
+            break;
+        case 2:
+            serialized = serializeVersion2(metadata);
+            break;
+        case 1:
+            serialized = serializeVersion1(metadata);
+            break;
+        default:
+            throw new IllegalArgumentException("Invalid format version " + formatVersion);
+        }
         if (log.isDebugEnabled()) {
-            log.debug("Serialized config: {}", s);
+            log.debug("Serialized with format {}: {}", formatVersion, Base64.getEncoder().encodeToString(serialized));
         }
-
-        return s.toString().getBytes(UTF_8);
+        return serialized;
     }
 
-    @VisibleForTesting
-    public LedgerMetadataFormat buildProtoFormat(LedgerMetadata metadata) {
+    private static byte[] serializeVersion3(LedgerMetadata metadata) throws IOException {
+        ByteArrayOutputStream os = new ByteArrayOutputStream();
+        writeHeader(os, 3);
```

Comment:
Use constant for '3' ?

Reply:
what would you name such a constant?

Reply:
Isn't it MAXIMUM_METADATA_FORMAT_VERSION ?

Or we can have 
METADATA_FORMAT_VERSION_3

Reply:
Uncomfortably close to ```#define THREE=3```. Will add them all the same.